### PR TITLE
[Relay, TOPI] add onehot op support

### PR DESCRIPTION
**Pull request by Beya2019**
add the onehot op support of tensorflow  in Relay and Topi. Currently onehot operator supports multiple on_values/off_value/depth, and 0 <= axis < data.shape.size(). The default value: on_value=1.0, off_value=0.0, axis=-1.

 @vinx13  would you be able to take a look?

### CODE REVIEW

File: python/tvm/relay/op/transform.py [hunk 1]
```diff
@@ -340,6 +340,46 @@ def arange(start, stop=None, step=None, dtype="float32"):
     return _make.arange(start, stop, step, dtype)
 
 
+def onehot(data, on_value=None, off_value=None, depth=None, axis=None):
```
Comment (vinx13) [inline python/tvm/relay/op/transform.py]: rename to `one_hot`

File: src/relay/op/tensor/transform.cc [hunk 2]
```diff
@@ -2482,5 +2482,105 @@ Examples::
...
+                Integer axis) {
+  auto attrs = make_node<OneHotAttrs>();
+  attrs->depth = std::move(depth);
+  attrs->axis = std::move(axis);
+  static const Op& op = Op::Get("onehot");
+  return CallNode::make(op, {data, on_value, off_value}, Attrs(attrs), {});
+}
+
+TVM_REGISTER_API("relay.op._make.onehot")
+.set_body_typed(MakeOneHot);
```
Comment (vinx13) [inline src/relay/op/tensor/transform.cc]: insert blank line below

File: src/relay/op/tensor/transform.cc [hunk 3]
```diff
@@ -2482,5 +2482,105 @@ Examples::
...
+.describe(R"code(Returns one-hot array within a given interval-depth.
+
+)code" TVM_ADD_FILELINE)
+.set_attrs_type_key("relay.attrs.OneHotAttrs")
+.set_num_inputs(3)
+.set_support_level(3)
+.add_type_rel("OneHot", OneHotRel)
+.set_attr<TOpPattern>("TOpPattern", kInjective)
+.set_attr<FTVMCompute>("FTVMCompute", OneHotCompute)
+.set_attr<AnyCodegenStrategy>("AnyCodegenStrategy", kVariableDimensions);
```
Comment (vinx13) [inline src/relay/op/tensor/transform.cc]: blank line below

File: tests/python/relay/test_op_level3.py [hunk 4]
```diff
@@ -685,6 +685,50 @@ def verify_gather_nd(xshape, yshape, y_data):
...
+def test_onehot():
+    def verify_onehot(src_shape, depth, on_value_data, off_value_data, axis=None):
+        data_dtype = "int32"
+        value_dtype = "float32"
+        shape_size = 1
+        for i in range(len(src_shape)):
+            shape_size = shape_size * src_shape[i]
+        input_data = np.arange(shape_size, dtype=data_dtype).reshape((src_shape))
+        input = relay.var("input", relay.TensorType(input_data.shape, data_dtype))
+        on_value = relay.var("on_value", relay.TensorType((1,), value_dtype))
```
Comment (vinx13) [inline tests/python/relay/test_op_level3.py]: use `relay.scalar_type` instead

File: topi/include/topi/transform.h [hunk 5]
```diff
@@ -660,6 +660,62 @@ inline Tensor take(const Tensor& a,
...
+* \param on_value The value to fill in output when indices[j] = i.
+* \param off_value The value to fill in output when indices[j] != i.
+* \param axis The axis over which to select values. By default,
+* the flattened input array is used.
+* \param name The name of the operation.
+* \param tag The tag to mark the operation.
+*
+* \return A Tensor whose op member is the take operation
+*/
+inline Tensor onehot(const Tensor& a,
```
Comment (vinx13) [inline topi/include/topi/transform.h]: rename to `one_hot`

File: topi/include/topi/transform.h [hunk 6]
```diff
@@ -660,6 +660,62 @@ inline Tensor take(const Tensor& a,
...
+        Array<Expr> real_indices;
+        for (int j = 0; j < axis; ++j) {
+            real_indices.push_back(indices[j]);
+        }
+        if (axis < input_shape) {
+          for (int j = axis + 1; j < static_cast<int>(indices.size()); ++j) {
+            real_indices.push_back(indices[j]);
+          }
+        }
+        Expr ret = tvm::if_then_else(a(real_indices) == indices[axis], on_value(0), off_value(0));
```
Comment (vinx13) [inline topi/include/topi/transform.h]: use `tvm::ir::Select::make` here, and use `on_value()` to access value of scalar type

File: topi/tests/python/test_topi_transform.py [hunk 7]
```diff
@@ -312,6 +312,45 @@ def check_device(device):
     for device in ["llvm", "opencl", "sdaccel", "aocl_sw_emu"]:
         check_device(device)
 
+
+def verify_onehot(src_shape, depth, on_value_data, off_value_data, axis=-1):
+    src_dtype = "int32"
+    value_dtype = "float32"
+    A = tvm.placeholder(shape=src_shape, dtype=src_dtype, name="indices")
+    on_value = tvm.placeholder(shape=(1,), dtype=value_dtype, name="on_value")
```
Comment (vinx13) [inline topi/tests/python/test_topi_transform.py]: use `tvm.placeholder(shape=())` to declare scalars

**Comment by jwfromm** [comment]
Worth noting that PR #3781 also implements this op and seems to have very similar changes.

**Comment by soiferj** [comment]
Yeah, I implemented basically the same thing in #3781. Doesn't matter to me which one is checked in, it would just be nice to get this functionality committed :)

**Comment by Beya2019** [comment]
Sorry, this is the first time I submitted the code on the TVM community, didn't notice the previous submission #3781


You can check in the better one. we all hope to improve and use this function as soon as possible.





At 2019-08-21 06:30:53, "Jon Soifer" <notifications@github.com> wrote:


Yeah, I implemented basically the same thing in #3781. Doesn't matter to me which one is checked in, it would just be nice to get this functionality committed :)

**Comment by soiferj** [comment]
My change to add OneHot was just checked in - feel free to give it a try!

**Comment by vinx13** [comment]
Closed this. You can send another PR if further changes are needed. Thanks for your contribution.

**Pull request closed without merging**